lib/pdk/config.rb

```diff
@@ -43,7 +49,26 @@ def user
     # @param filter [String] Only resolve setting names which match the filter. See PDK::Config::Namespace.be_resolved? for matching rules
     # @return [Hash{String => Object}] All resolved settings for example {'user.module_defaults.author' => 'johndoe'}
     def resolve(filter = nil)
-      user.resolve(filter)
+      self.system.resolve(filter).merge(user.resolve(filter)) # rubocop:disable Style/RedundantSelf Use an explicit call so as to not confuse with Kernel.system
+    end
+
+    def get(*name)
+      return nil if name.nil? || !name.is_a?(Array) || name.empty?
+      case name[0]
+      when 'user'
+        traverse_object(user, *name[1..-1])
+      when 'system'
+        traverse_object(self.system, *name[1..-1]) # rubocop:disable Style/RedundantSelf Use an explicit call so as to not confuse with Kernel.system
+      else
+        nil
+      end
+    end
+
+    # Gets a named setting using precedence from the user and system levels
+    # Note that name must NOT include user or system prefix
+    def pdk_setting(*name)
+      value = get(*['user'].concat(name))
+      value.nil? ? get(*['system'].concat(name)) : value
```

Comment:
I feel like we are mostly going to want the "lookup using precedence" behavior that is currently implemented in `pdk_setting()`, so maybe what is currently `get()` should be like `get_within_scope(scope, *name)` and then the "lookup using precedence" can just become `get()`?

Reply:
I disagree here, but I believe I can the point your trying to make here. And I guess it depends on the consumers point of view.

In the future, we'll have at least one addition setting level, `project.`, or even multiple different ones (`control_repo.` or `module.`).

So I see

`PDK::Config.get(...)` to be the most specific form. "Get me this setting name"

`PDK::Config.pdk_setting` to be a helper, to stop copying the precedence logic.  I chose `pdk_setting` so it would (hopefully) be obvious that this is only a PDK setting, not a Project setting.

I do concede that `PDK::Config.pdk_setting` seems like a redundant name.

As for your suggestion:

`get_within_scope(scope, *name)`  makes perfect sense

but `get()` only returning `user` and `system` level settings may be a confusing method name.

Reply:
That said, I do like that we both agree on the layout of the methods, just the method names need some discussion 👍

Reply:
@scotje You happy for merge are would prefer a name change?

Reply:
Doesn't it seem possible people will want to override a "PDK" setting at the `project` scope at some point as well? Or is it clear that the system/user settings will never overlap with project settings?

It seems like what we need for a general interface is:

- get a setting based on default precedence rules (I think that would be `project > user > system`?)
- get a setting from a specific precedence level
- (maybe?) get a setting from a specific subset of precedence levels (for when the caller has a special reason for bypassing normal precedence)

So maybe it's just that `get_within_scope` should also accept a list for the `scope` parameter? Or we could have a `get_within_scopes`?

Reply:
Okay, I've basically implemented what you stated. I've gone with:

* `.get`  This gets a fully named setting e.g. `user.foo.bar`.  No scope precedence or anything

* `.get_within_scopes` : This gets a setting name in zero or more specified scopes (in order).  If no scopes are given (nil) then it searches in the default scopes in the default order.

Which ends up with:
* For the bulk of calls within /lib/cli/get/config `.get` is probably the best
* For the bulk of calls within /lib/* `.get_within_scopes` is probably the best
* For critical calls like analytics information, it can be restricted to `.get_within_scopes('....', ['user', 'system'])`.  Later on, I may add a shortcut method (`def critical_scopes; ['user', 'system`].freeze; end;`) if I find that we use it too often.